### ISSUE TITLE
Cyclic reference involving class for Java dependency

### ISSUE THREAD

**Issue by cchantep** [lampepfl/dotty#11828]
## Compiler version

3.0.0-RC1

## Minimized code

```Scala
import acolyte.jdbc.RowLists

(RowLists.stringList :+ v).asResult
```

See:
- [source](https://github.com/cchantep/acolyte/blob/master/jdbc-scala/src/main/scala/AcolyteDSL.scala#L376)
- [Java class](https://oss.sonatype.org/service/local/repositories/releases/archive/org/eu/acolyte/jdbc-driver/1.0.54/jdbc-driver-1.0.54-javadoc.jar/!/acolyte/jdbc/RowLists.html)

## Output

```
exception caught when loading class Impl: Cyclic reference involving class Impl
[error] -- [E046] Cyclic Error: /Users/cchantep/Projects/acolyte/jdbc-scala/src/main/scala/AcolyteDSL.scala:377:5 
[error] 377 |    (RowLists.stringList :+ v).asResult
[error]     |     ^
```

## Expectation

Should stay compatible with Java classes.

**Comment by griggt** [comment]
Self-contained:

#### RowList1.java
```java
public class RowList1<A, U extends RowList1<A, ?>> {
    public static final class Impl<A> extends RowList1<A, Impl<A>> {}
}
```

#### RowLists.java
```java
public final class RowLists {
    public static RowList1.Impl<String> stringList() {
        throw new UnsupportedOperationException();
    }
}
```

#### Test.scala

```scala
object Test:
  val test = RowLists.stringList
```

```
$ javac -version
javac 1.8.0_282

$ dotc -version
Scala compiler version 3.0.0-RC1 -- Copyright 2002-2021, LAMP/EPFL

$ javac RowList1.java RowLists.java && dotc Test.scala 
exception caught when loading class Impl: Cyclic reference involving class Impl
-- [E046] Cyclic Error: Test.scala:2:13 ----------------------------------------
2 |  val test = RowLists.stringList
  |             ^
  |             Cyclic reference involving class Impl
```

**Comment by griggt** [comment]
This seems to have already been fixed by #11471, could you try a recently nightly and confirm?

**Comment by cchantep** [comment]
> This seems to have already been fixed by #11471, could you try a recent nightly and confirm?

Which version?

**Comment by cchantep** [comment]
Seems ok, in 3.0.0-RC2, thanks.
